harness/determined/cli/sso.py

```diff
@@ -73,17 +77,33 @@ def sso(parsed_args: Namespace) -> None:
         matched_provider = matching_providers[0]
 
     sso_url = matched_provider["sso_url"] + "?relayState=cli"
-    webbrowser.open(sso_url)
-    print(
-        "Your browser should open and prompt you to sign on;"
-        " if it did not, please visit {}".format(sso_url)
-    )
 
-    with HTTPServer(
-        ("localhost", CLI_REDIRECT_PORT),
-        make_handler(parsed_args.master, lambda code: sys.exit(code)),
-    ) as httpd:
-        httpd.serve_forever()
+    if not parsed_args.headless and webbrowser.open(sso_url):
+        print(
+            "Your browser should open and prompt you to sign on;"
+            " if it did not, please visit {}".format(sso_url)
+        )
+        with HTTPServer(
+            ("localhost", CLI_REDIRECT_PORT),
+            make_handler(parsed_args.master, lambda code: sys.exit(code)),
+        ) as httpd:
+            httpd.serve_forever()
+    else:
+        example_url = f"Example: 'http://localhost:{CLI_REDIRECT_PORT}/?token=v2.public.[long_str]'"
+        print(
+            "Failed to open Web Browser. Assuming headless CLI.\n"
+            f"Please open this URL in your browser: '{sso_url}'\n"
+            "After authenticating, copy/paste the localhost URL "
+            f"from your browser into the prompt.\n{example_url}"
+        )
```

Comment:
this message will print ("failed to open web browser") even when `--headless` is passed, you'll need a more elaborate control flow here:

```python
if not parsed_args.headless
    if webbrowser.open(sso_url):
        ...
        return httpd.serve_forever()
    print("Failed to open Web Browser. Falling back to --headless CLI mode.\n"

# --headless or webbrowser failed
print(f"Please open this URL in your browser ...")
token = None
while not Token:
    ...
```

Reply:
Thanks again! Great catch!